monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2014-2018 by The Monix Project Developers.
+ * See the project homepage at: https://monix.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.operators
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.locks.ReentrantLock
```

Comment:
I'm not too familiar with Scala.js - should I only rely on `monix` primitives / implement something new if needed?

Reply:
For `ConcurrentLinkedQueue` I'm pretty sure we have an alias or wrapper somewhere in `monix.execution` that you can use.

In the case of `ReentrantLock`, I have no idea what the Scala.js version is doing. Should be a dummy. I guess that if the tests pass, then it's not bad. Was surprised to see it in shared code, guess Scala.js has evolved its Java library.

So you can leave it, but was just mentioning that in case you introduce it in shared code, then you need to understand what the Scala.js implementation is doing, to verify that it's OK.

Reply:
Leaving note about Scala.JS:

Reentrant lock:
https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/concurrent/locks/ReentrantLock.scala

So it just returns `true` for `tryLock` every time, sounds safe to use.

ConcurrentLinkedQueue:
https://github.com/scala-js/scala-js/blob/master/javalib/src/main/scala/java/util/concurrent/ConcurrentLinkedQueue.scala

This is a bit more complicated but glancing through the code it looks like ordinary `Queue`. 
I've found [Monix's ConcurrentQueue](https://github.com/monix/monix/blob/master/monix-reactive/jvm/src/main/scala/monix/reactive/observers/buffers/ConcurrentQueue.scala) but it's `private[buffer]`.

BTW @alexandru  

I'm trying to see how I can "merge" or remove any of those structures:
1. `AsyncSemaphore` - I need to limit incoming tasks.
2. `BufferedSubscriber` - I call `subscriber.onNext` concurrently.
3. `ConcurrentLinkedQueue` - order elements, need synchronized access because when one thread is checking `Queue` for head completion, new elements can arrive and they should be added to the queue.
4. `ReentrantLock` - I use it to ensure that only one thread at the time checks `Queue`, I only use `tryLock` so it shouldn't be big overhead


I'm trying to figure out if I can get rid of `BufferedSubscriber` since I call `onNext` only from one place but no luck so far - it's failing few tests.

Other than that I have only desparate idea which is implementing custom semi-bounded thread-safe  `Queue` which could track completion of its items and only allow up to N not-completed elements. Then backpressuring on it instead of `AsyncSemaphore` :D But even if that makes any sense I don't think I have capacity to do it soon and we probably want this operator in Monix 3.0.0 - a lot of people are requesting this.